tests/integration/test_query.py

```diff
@@ -0,0 +1,220 @@
+    v = VectorQuery(
+        [0.1, 0.1, 0.5],
+        "user_embedding",
+        return_fields=["user", "credit_score", "age", "job"],
+        hybrid_filter=t,
+    )
+
+    results = index.search(v.query, query_params=v.params)
+    assert len(results.docs) == 4
+
+
+def test_simple_numeric_filter(index):
+    # (@age:[18 101])=>[KNN 10 @user_embedding $vector AS vector_distance]
+    n = NumericFilter("age", 18, 100)
+    v = VectorQuery(
+        [0.1, 0.1, 0.5],
+        "user_embedding",
+        return_fields=["user", "credit_score", "age", "job"],
+        hybrid_filter=n,
+    )
+
+    results = index.search(v.query, query_params=v.params)
+    assert len(results.docs) == 4
+
+
+def test_numeric_filter_exclusive(index):
+    n = NumericFilter("age", 18, 100, min_exclusive=True)
+    v = VectorQuery(
+        [0.1, 0.1, 0.5],
+        "user_embedding",
+        return_fields=["user", "credit_score", "age", "job"],
+        hybrid_filter=n,
+    )
+
+    results = index.search(v.query, query_params=v.params)
+    assert len(results.docs) == 3
+
+    n_both_exclusive = NumericFilter(
+        "age", 18, 100, min_exclusive=True, max_exclusive=True
+    )
+    v.set_filter(n_both_exclusive)
+    results = index.search(v.query, query_params=v.params)
+    assert len(results.docs) == 2
+
+
+def test_combinations(index):
+    # (@age:[18 100] @credit_score:{high})=>[KNN 10 @user_embedding $vector AS vector_distance]
+    t = TagFilter("credit_score", "high")
+    n = NumericFilter("age", 18, 100)
+    t += n
+    v = VectorQuery(
+        [0.1, 0.1, 0.5],
+        "user_embedding",
+        return_fields=["user", "credit_score", "age", "job"],
+        hybrid_filter=t,
+    )
+
+    results = index.search(v.query, query_params=v.params)
+    for doc in results.docs:
+        assert doc.credit_score == "high"
+        assert 18 <= int(doc.age) <= 100
+    assert len(results.docs) == 3
+
+    # (@credit_score:{high} -@age:[18 100])=>[KNN 10 @user_embedding $vector AS vector_distance]
+    t = TagFilter("credit_score", "high")
+    n = NumericFilter("age", 18, 100)
+    t -= n
+    v.set_filter(t)
+
+    results = index.search(v.query, query_params=v.params)
+    for doc in results.docs:
+        assert doc.credit_score == "high"
+        assert int(doc.age) not in range(18, 101)
+    assert len(results.docs) == 1
+
+    # (@credit_score:{high} | @age:[18 100])=>[KNN 10 @user_embedding $vector AS vector_distance]
+    t = TagFilter("credit_score", "high")
+    n = NumericFilter("age", 18, 100)
+    t &= n
+    v.set_filter(t)
+
+    results = index.search(v.query, query_params=v.params)
+    for doc in results.docs:
+        assert (doc.credit_score == "high") or (18 <= int(doc.age) <= 100)
+    assert len(results.docs) == 5
+
+    # (@credit_score:{high} ~@age:[18 100])=>[KNN 10 @user_embedding $vector AS vector_distance]
+    t = TagFilter("credit_score", "high")
+    n = NumericFilter("age", 18, 100)
+    t ^= n
+    v.set_filter(t)
+
```

Comment:
only other input on tests would be to test some more tricky character scenarios where we'd have to properly escape them to make it work

Reply:
Good call. Examples? Feel like you know more about this than I do.